addons/account/models/account.py

```diff
@@ -253,6 +256,24 @@ def _check_currency(self):
             if self.default_debit_account_id and not self.default_debit_account_id.currency_id.id == self.currency_id.id:
                 raise UserError(_('Configuration error!\nThe currency of the journal should be the same than the default debit account.'))
 
+    @api.one
+    @api.constrains('type', 'bank_account_id')
+    def _check_bank_account(self):
+        if self.type == 'bank':
+            # A journal of type bank is necessarily composed of a bank account
+            if not self.bank_account_id:
+                raise ValidationError(_('A bank journal cannot be created without a bank account.'))
+            # DISCLAIMER : There is a painful duality with bank accounts. Either they belong to a customer/supplier,
+            # in which case their partner_id is the customer/supplier and company_id is for multicompany settings.
+            # Or they are part of a journal, in this case they must have the same company_id as the journal and the
+            # partner_id should be left empty. However, for backwards compatibility, we set partner_id to company_id.partner_id
+            if self.bank_account_id.company_id != self.company_id:
+                raise ValidationError(_('The bank account of a bank journal must belong to the same company.'))
+            if not self.bank_account_id.partner_id:
```

Comment:
I think it's a very bad practice to set a value during a constraint check, I would remove the:
    self.bank_account_id.partner_id = self.company_id.partner_id
and, eventually, raise an exception instead.